Framework/API/src/SpectrumInfo.cpp

```diff
@@ -176,8 +176,15 @@ const Geometry::IDetector &SpectrumInfo::getDetector(const size_t index) const {
     // Else need to construct a DetectorGroup and use that
     std::vector<boost::shared_ptr<const Geometry::IDetector>> det_ptrs;
     for (const auto &id : dets) {
-      const auto detIndex = m_detectorInfo.indexOf(id);
-      det_ptrs.push_back(m_detectorInfo.getDetectorPtr(detIndex));
+      try {
+        const auto detIndex = m_detectorInfo.indexOf(id);
+        det_ptrs.push_back(m_detectorInfo.getDetectorPtr(detIndex));
+      } catch (std::out_of_range) {
+        // Workspaces can contain invalid detector IDs. Those IDs will be
+        // silently ignored here until this is fixed. Some valid IDs will exist
+        // if hasDetectors or hasUniqueDetectors has returned true, but there
+        // could still be invalid IDs.
+      }
```

Comment:
You cans see the system tests that failed here - http://builds.mantidproject.org/job/pull_requests-rhel7/10305/.

In the case of the `LoadTest` failures it is caused by running something like `Load("TSC15352+15353.raw)`. There is something going wrong in the loader, so one spectra has a spectrum number of `-1` and is associated with mostly invalid detector IDs, but some valid ones are still in there.

I did not look much into the SNSConvertToMD*, but something similar seems to happen, though the spectrum numbers are at least valid.

I do not know how expensive the try catch would be compared with other approaches. I did also try iterating over the detectors in `getDetector()` to check if they were valid, but this seemed messy and even more expensive.

The better option may be to get the bad loaders fixed now, but it may take some time to work through fixing them.